# Introduction to DataFrames

In [ ]:
Introduction

    The pandas module has been developed to provide Python with the tools necessary to manipulate and analyze large volumes of data.

    Pandas introduces the DataFrame class, an array-like data structure that offers more advanced data manipulation and exploration than NumPy arrays.

    The main features of pandas are:

            data recovery from files (CSV, Excel tables, etc.)

            handling this data (deletion / addition, modification, statistical visualization, etc.).

    This notebook aims at:

            Understanding the format of a DataFrame.

            Creating a first Dataframe.

            Carrying out a first exploration of a dataset using the DataFrame class.

    (a) Import the pandas module under the name pd.


In [ ]:
# Insert your code here
import pandas as pd

## Format of a DataFrame

In [ ]:
1. Format of a DataFrame

    A DataFrame is in the form of a matrix whose rows and columns each have an index. Typically, columns are indexed by name and rows by unique identifiers.

    A DataFrame is used to store a database. The different entries in the database (individuals, animals, objects, etc.) are the different lines and their features are the different columns:
    	Name 	Gender 	Height 	Age
    0 	Robert 	M 	174 	23
    1 	Mark 	M 	182 	40
    2 	Aline 	F 	169 	56

            The DataFrame above groups together information on 3 individuals: the DataFrame therefore has 3 lines.

            For each of these individuals, there are 4 variables (name, gender, height and age) : therefore, the DataFrame has 4 columns.

    The column containing the numbering of the lines is called the index and is not managed in the same way as other columns of the DataFrame.

    The index can be set by default (will follow the row numbering), defined with one (or several) of the columns of the DataFrame or even defined with a list that we specify.

    Example: Default indexing (line numbering), you don't have to specify anything :
    	Name 	Gender 	Height 	Age
    0 	Robert 	M 	174 	23
    1 	Mark 	M 	182 	40
    2 	Aline 	F 	169 	56

    Example: Indexing by the column 'Name':
    	Gender 	Height 	Age
    Robert 	M 	174 	23
    Mark 	M 	182 	40
    Aline 	F 	169 	56

    Example: Indexing by the list ['person_1', 'person_2', 'person_3']:
    	Name 	Gender 	Height 	Age
    person_1 	Robert 	M 	174 	23
    person_2 	Mark 	M 	182 	40
    person_3 	Aline 	F 	169 	56

    We will detail later how to define the index when creating a DataFrame.

    The DataFrame class has several advantages over aNumpy array:

            Visually, a DataFrame is much more readable thanks to more explicit column and row indexing.

            Within the same column the elements are of the same type but from one column to another, the type of the elements may vary, which is not the case of Numpy arrays which only support data of the same type.

            The DataFrame class contains more methods for handling and preprocessing databases, while NumPy specializes instead in optimized computation.



## Creation of a DataFrame: from a NumPy array

In [ ]:
2. Creation of a DataFrame: from a NumPy array

    It is possible to directly create a DataFrame from a NumPy array using the DataFrame() constructor. The disadvantage of this method is that it is not very practical and the data type is necessarily the same for all the columns.

    Let's take a closer look at the header of this constructor.

    pd.DataFrame(data, index, columns, ...)

            The data parameter contains the data to be formatted (NumPy array, list, dictionary or another DataFrame).

            The index parameter, if specified, must be a list containing the indices of the entries.

            The columns parameter, if specified, must be a list containing the name of the columns.

    For other parameters, you can consult the Python documentation.

    Example:

    # Creation of a NumPy array with 3 rows and 4 columns
    array = np.array ([[1, 2, 3, 4],
                       [5, 6, 7, 8],
                       [9, 10, 11, 12]])

    # Instantiation of a DataFrame
    df = pd.DataFrame (data = array, # The data to format
                       index = ['i_1', 'i_2', 'i_3'], # The indices of each entry
                       columns = ['A', 'B', 'C', 'D']) # The name of the columns

    This produces the following DataFrame:
    	A 	B 	C 	D
    i_1 	1 	2 	3 	4
    i_2 	5 	6 	7 	8
    i_3 	9 	10 	11 	12

## Creation of a DataFrame: from a dictionary

In [ ]:
3. Creation of a DataFrame: from a dictionary

    Another way to create a DataFrame is to use a dictionary. Thanks to this technique, the columns can be of different type and their names are already given when creating the DataFrame.

    Example:

    # Creation of a dictionary
    dictionary = {'A': [1, 5, 9],
                  'B': [2, 6, 10],
                  'C': [3, 7, 11],
                  'D': [4, 8, 12]}

    # Instantiation of a DataFrame
    df = pd.DataFrame (data = dictionary,
                       index = ['i_1', 'i_2', 'i_3'])

    This produces the same DataFrame as before:
    	A 	B 	C 	D
    i_1 	1 	2 	3 	4
    i_2 	5 	6 	7 	8
    i_3 	9 	10 	11 	12

The manager of a grocery store has the following stock of food products:

    100 jars of honey with an expiration date of 08/10/2025 and worth €2 each.

    55 packets of flour expiring on 09/25/2024 each costing € 3.

    1800 bottles of wine costing € 10 per unit and expiring on 10/15/2023.

    (a) From a dictionary, create and display a DataFrame df that contains for each product:

            Its name.
            Its expiration date.
            Its quantity.
            Its price per unit.

You will choose relevant column names and the index will be the default one (in this case we do not specify the index parameter).

In [42]:
# Insert your Code

dictionary = {"Product"          : ['honey', 'flour', 'wine'],
              "Expiration date"  : ['10/08/2025', '25/09/2024', '15/10/2023'],
              "Quantity"         : [100, 55, 1800], 
              "Price per unit"   : [2, 3, 10]}

df = pd.DataFrame(dictionary)

df

Product Expiration date  Quantity  Price per unit
0   honey      10/08/2025       100               2
1   flour      25/09/2024        55               3
2    wine      15/10/2023      1800              10

## Creation of a DataFrame: from a data file

In [ ]:
4. Creation of a DataFrame: from a data file

    Most often a DataFrame is created directly from a file containing the data of interest. The file's format can be a CSV, Excel, txt, etc.

    The most common format is the CSV format, which stands for Comma-Separated Values and denotes a spreadsheet-like file whose values are separated by commas.

    Here is an example:

    A, B, C, D,
    1, 2, 3, 4,
    5, 6, 7, 8,
    9, 10, 11, 12

    In this format:

            The first line contains the name of the columns, but sometimes the name of the columns is not filled in.

            Each line corresponds to an entry in the database.

            The values are separated by a separator character. In this example, it is ',' but it could be a ';'.

    To import the data into a DataFrame, we need to use the read_csv function of pandas whose header is as follows:

    pd.read_csv(filepath_or_buffer, sep = ',', header = 0, index_col = 0 ...)

    The essential arguments of the pd.read_csv function to know are:

            filepath_or_buffer: The path of the .csv file relative to the execution environment.
                If the file is in the same folder as the Python environment, just fill in the name of the file.
                This path must be entered in the form of character string.

            sep: The character used in the .csv file to to separate the different columns.
                This argument must be specified as character.

            header: The number of the row that contains the names of the columns.
                If for example the column names are entered in the first line of the .csv file, then we must specify header = 0.
                If the names are not included, we will put header = None.

            index_col: The name or number of the column containing the indices of the database.
                If the database entries are indexed by the first column, you will need to fill in index_col = 0.
                Alternatively, if the entries are indexed by a column which bears the name "Id", we can specify index_col = "Id".

    This function will return an object of type DataFrame which contains all the data of the file.

    (a) Load the data contained in the file transactions.csv into aDataFrame named transactions:

            The file is located in the same folder as the environment of this notebook.
            Columns are separated by commas.
            The names of the columns are in the first line of the file.
            The rows of the database are indexed by the "transaction_id" column which is also the first column.



In [ ]:
# insert your code

# You can directly specify the name of the column containing the indices

transactions = pd.read_csv(filepath_or_buffer = 'transactions.csv', # file path
                           sep = ',',                               # character separating values
                           header = 0,                              # number of the row containing column names
                           index_col = 'transaction_id')            # name of the column that indexes the entries


# You can also directly enter the number of the column that indexes the entries

transactions = pd.read_csv(filepath_or_buffer = 'transactions.csv',
                           sep = ',',
                           header = 0,
                           index_col = 0) # number of the column that indexes the entries

# We loaded the transactions.csv file in theDataFrame transactions which gathers a history of transactions carried out between 2011 and 2014. 
#In the next section, we will study this dataset.

## First exploration of a dataset using the DataFrame class

In [ ]:
5. First exploration of a dataset using the DataFrame class

    The rest of this notebook briefly presents the main methods of the DataFrame class which will allow us to do a quick analysis of our data set, that is:

            Having a brief overview of the data (head method,columns and shape attributes).

            Selecting values in the DataFrame (loc and iloc methods).

            Carrying out a quick statistical study of our data (describe and value _counts methods)

    As a reminder, to apply a method to an object in Python (such as a DataFrame for example), you must add the method as a suffix of the object. Example: my_object.my_method()


## Visualization of a DataFrame: head method, columns and shape attributes

In [ ]:
6. Visualization of a DataFrame: head method, columns and shape attributes

        It is possible to have a preview of a dataset by displaying only the first lines of the DataFrame.

    For that, we must use the head() method, specifying as an argument the number of lines that we want to display (by default 5).

    It is also possible to preview the last lines using the tail() method which is applied in the same way:

    # Display of the first 10 lines of my_dataframe
    my_dataframe.head(10)

    (a) Display the first 20 lines of the transactions DataFrame.


In [ ]:
# Insert your code here
transactions.head(20)

In [ ]:
# (b) Display the last 10 lines of the transactions DataFrame.

In [ ]:
# Insert your code here
transactions.tail(10)

In [ ]:
     	A 	B 	C 	D
    i_ 1 	1 	2 	3 	4
    i _2 	5 	6 	7 	8
    i_ 3 	9 	10 	11 	12

    # Display of df DataFrame columns
    print(df.columns)
    >>> ['A', 'B', 'C', 'D']

    The list of the column names can be used to iterate over the columns of a DataFrame within a loop.

    It can be interesting to know how many transactions (rows) and how many features (columns) the dataset contains.

    For this we will use the shape attribute of the DataFrame class which displays the dimensions of our DataFrame in the form of a tuple (number of rows, number of columns):

    # Display the dimensions of df
    print (df.shape)
    >>> (3,4)

    (c) Display the dimensions of the DataFrame transactions as well as the name of the 5th column. Remember that in Python the indices start from 0.


In [ ]:
# Insert your code here
print(transactions.shape)

transactions.columns[4]

## Selecting columns from a DataFrame

In [ ]:
7. Selecting columns from a DataFrame

    Extracting columns from a DataFrame is almost identical to extracting data from a dictionary.

    To extract a column from a DataFrame, all we have to do is enter between brackets the name of the column to extract. To extract several columns, we must enter between brackets the list of the names of the columns to extract:

    # Display of the 'cust_id' column
    print(transactions['cust_id'])

    # Extraction of 'cust_id' and 'Qty' columns from transactions
    cust_id_qty = transactions[["cust_id", "Qty"]]

    cust_id_qty is a new DataFrame containing only the 'cust_id' and 'Qty' columns.

    The display of the first 3 lines of cust_id_qty yields:



    transactions_id 	cust_id 	Qty
    80712190438 	270351 	-5
    29258453508 	270384 	-5
    51750724947 	273420 	-2

    When we prepare a dataset for later use, it is better to separate the categorical variables from the quantitative variables:

            A categorical variable is a variable that takes only a finite number of modalities.

            The categorical variables of the DataFrame transactions are: ['cust_id', 'tran_date', 'prod_subcat_code', 'prod_cat_code', 'Store_type'].

            A quantitative variable is a variable that measures a quantity that can take an infinite number of values.

            The quantitative variables of transactions are: ['Qty', 'Rate', 'Tax', 'total_amt'].

    This distinction is made because some basic operations like calculating an average only make sense for quantitative variables.

    (a) In a DataFrame named cat_vars, store the categorical variables of transactions.

    (b) In a DataFrame named num_vars, store the quantitative variables of transactions.

    (c) Display the first 5 lines of each DataFrame.



In [ ]:
# insert your code

# Extraction of categorical variables
cat_var_names = ['cust_id', 'tran_date', 'prod_subcat_code', 'prod_cat_code' , 'Store_type']
cat_vars = transactions[cat_var_names]

# Extraction of quantitative variables
num_var_names = ['Qty', 'Rate', 'Tax', 'total_amt']
num_vars = transactions[num_var_names]

# Display of the first 5 lines of each DataFrame
print ("Categorical variables: \n")
print (cat_vars.head(), "\n \n")

print ("Quantitative variables: \n")
print (num_vars.head())

## Selecting rows of a DataFrame: loc and iloc methods

In [ ]:
8. Selecting rows of a DataFrame: loc and iloc methods

    To extract one or more rows from a DataFrame, we use the loc method. loc is a very special type of method because the arguments are filled in between square brackets and not between parentheses. Using this method is very similar to indexing lists.

    In order to retrieve the line of index i of a DataFrame, all we have to do is enter i as an argument of the loc method:

    # We retrieve the line of index 80712190438 of the num_vars DataFrame
    print(num_vars.loc[80712190438])

    >>                 Rate    Tax  total_amt
    >> transaction_id                         
    >> 80712190438    -772.0  405.3    -4265.3
    >> 80712190438     772.0  405.3     4265.3

    In order to retrieve several rows, we can either:

            Enter a list of indices.

            Enter a slice by specifying the start and end indices of the slice. To use slicing with loc, the indices must be unique, which is not the case for transactions.

    # We retrieve the rows at indices 80712190438, 29258453508 and 51750724947 from the transactions DataFrame
    transactions.loc[[80712190438, 29258453508, 51750724947]]

    loc can also take a column or list of columns as an argument in order to refine the data extraction:

    # We extract the columns 'Tax' and 'total_amt' from the rows at index 80712190438 and 29258453508
    transactions.loc[[80712190438, 29258453508], ['Tax', 'total_amt']]

    This instruction produces the following DataFrame:



    transaction_id 	Tax 	total_amt
    80712190438 	405.300 	-4265.300
    80712190438 	405.300 	4265.300
    29258453508 	785.925 	-8270.925
    29258453508 	785.925 	8270.925

    The iloc method is used to index a DataFrame exactly like a numpy array, that is to say by only filling in the numeric indices of the rows and columns. This allows the use of slicing without constraint:

    # Extraction of the first 4 rows and the first 3 columns of transactions
    transactions.iloc[0:4, 0:3]

    This instruction produces the following DataFrame:



    transaction_id 	cust_id 	tran_date 	prod_subcat_code
    80712190438 	270351 	28-02-2014 	1.0
    29258453508 	270384 	27-02-2014 	5.0
    51750724947 	273420 	24-02-2014 	6.0
    93274880719 	271509 	24-02-2014 	11.0

    If the row indexing is the one by default (row numbering), the loc and iloc methods are equivalent.





## Conditional indexing of a DataFrame

In [ ]:
9. Conditional indexing of a DataFrame

    As with Numpy arrays, we can use conditional indexing to extract rows from a Dataframe that meet a given condition.

    In the following illustration, we select the rows of the DataFrame df for which the column col 2 is equal to 3.


    There are two syntaxes for conditionally indexing a DataFrame:


    # We select the rows of the DataFrame df for which the column 'col 2' is equal to 3.
    df[df['col 2'] == 3]

    df.loc[df['col 2'] == 3]

    If we want to assign a new value to these entries, we must absolutely use the loc method.

    Indeed, indexing with the syntax df[df['col 2'] == 3] only returns a copy of these entries and does not provide access the memory location where the data is located.

The manager of the transactions listed in the transactions DataFrame wishes to have access to the identifiers of customers who have made an online purchase (i.e. in a "e-Shop") as well as the date of the corresponding transaction.

We have the following information about the columns of transactions:
Column name 	Description
'cust_id' 	The identifier of the customer
'Store_type' 	The type of store where the transaction took place
'tran_date' 	The date of the transaction

    (a) In a DataFrame named transactions_eshop, store the transactions that took place in an "e-Shop" type store.

    (b) In another DataFrame named transactions_id_date, store the customer identifiers and the transaction date of the transactions_eshop DataFrame.

    (c) Display the first 5 rows of transactions_id_date.

In [ ]:
# insert your code

# Creation of transactions_eshop by conditional indexing
transactions_eshop = transactions.loc[transactions['Store_type'] == 'e-Shop']

# Extraction of the 'cust_ id' and 'tran _date' columns
transactions_id_date = transactions_eshop[['cust_id', 'tran_date']]

# Display of the first 5 lines of transactions_id_date
transactions_id_date.head()

In [ ]:
Now, the manager would like to have access to the transactions carried out by the client whose identifier is 268819.

    (d) In a DataFrame named transactions_client_268819, store all transactions with client identifier 268819.

    (e) A column in a DataFrame can be iterated over with a loop exactly like a list (for value in df['column']:). Using a for loop on the 'total_amt' column, compute and display the total transaction amount for the client with identifier 268819.



In [ ]:
# Insert your code here

# Extraction of the transactions ofthe customer which identifier is 268819
transactions_client_268819 = transactions[transactions['cust_id'] == 268819]


# Computation of the total amount of transactions
total = 0

# For each amount in the column 'total_amt'
for amount in transactions_client_268819['total_amt']:
    # We sum the amounts
    total += amount
    
print(total)

# Second Way
transactions.loc[transactions.cust_id == 268819]['total_amt'].sum()

## Quick statistical study of the data in a DataFrame.

In [ ]:
10. Quick statistical study of the data in a DataFrame.

    The describe method of a DataFrame returns a summary of the descriptive statistics (min, max, mean, quantiles,...) of its quantitative variables. It is therefore a very useful tool for a first visualisation of the type and distribution of these variables.

    To analyse the categorical variables, it is recommended to start by using the value_counts method which returns the number of occurrences for each modality of these variables. The value_counts method cannot be used directly on a DataFrame but only on the columns of the DataFrame which are objects of the pd.Series class.

    (a) Use the describe method of the DataFrame transactions.

    (b) The quantitative variables of transactions are 'Qty', 'Rate', 'Tax' and total_amt'. By default, are the statistics produced by the describe method only computed on the quantitative variables?

    (c) Display the number of occurrences of each modality of the Store_type column using the value_counts method.



In [ ]:
# Insert your code here

transactions.describe()

transactions['Store_type'].value_counts()

In [ ]:


    The describe method computed statistics on the variables cust_id, prod_subcat_code and prod_cat_code while these are categorical variables.

    Of course, these statistics make no sense. The describe method has treated these variables as quantitative because the modalities they take are of numerical type.

    This is why it is necessary to pay attention to the results returned by the describe method and always take a step back to remember what the variables are reflecting.

    The manager wishes to make a quick report on the characteristics of the transactions DataFrame: in particular, he wants to know the average amount spent as well as the maximum quantity purchased.

    (d) What is the average total amount spent? We are interested in the 'total_amt' column of transactions.

    (e) What is the maximum quantity purchased? We will look at the 'Qty' column of transactions.



In [ ]:
# Insert your code here

print('average total amount spent :', transactions['total_amt'].mean())
print('maximum quantity purchased :', transactions['Qty'].max())

transactions.describe()

# Applying the describe method to the transactions DataFrame
transactions.describe()

# The average total amount spent is €2109.
# The maximum quantity purchased is 5. 

In [ ]:
Some transactions have negative amounts.

These are transactions that have been cancelled and refunded to the client. These amounts will disrupt the distribution of the amounts which gives us bad estimates of the mean and quantiles of the variable total_amt.

    (f) What is the average amount of transactions with positive amounts?



In [ ]:
# insert your code here

transactions[transactions['total_amt'] > 0].describe()

# the average amount of transactions with positive amounts is worth €2608 which is
# €500 more than we had before.

## Conclusion and recap

In [ ]:
11. Conclusion and recap

    The DataFrame class of the pandas module will be your favorite data structure when exploring, analysing and processing datasets and databases.

    In this brief introduction, you have learned to:

            Create a DataFrame from a numpy array and a dictionary using the pd.DataFrame constructor.

            Create a DataFrame from a .csv file using the pd.read_csv function.

            Display the first and last lines of a DataFrame using the head and tail methods.

            Select one or more columns of a DataFrame by entering their names in square brackets as in a dictionary.

            Select one or more lines of a DataFrame by filling in their index using the loc and iloc methods.

            Select the lines of a DataFrame that meet a specific condition using conditional indexing.

            Perform a quick statistical study of the quantitative variables of a DataFrame using the describe method.

    The dataset transactions we used is very clean. The variables are cleanly filled in and do not contain any missing value. In practice, this is rarely the case. This is why in the following notebook we will see how to clean datasets with pandas.



In [28]:
import pandas as pd
df= pd.read_csv('transactions.csv', index_col='transaction_id')
df['\ttran_date'] = df['\ttran_date'].apply(lambda x : x.replace('\t', ''))
df.rename(columns= lambda x : x.replace('\t', ''), inplace=True)
df

cust_id   tran_date  prod_subcat_code  prod_cat_code  qty  \
transaction_id                                                              
80712190438      270351  28-02-2014                 1              1   -5   
29258453508      270384  27-02-2014                 5              3   -5   
51750724947      273420  24-02-2014                 6              5   -2   
93274880719      271509  24-02-2014                11              6   -3   
51750724947      273420  23-02-2014                 6              5   -2   

                  rate      tax  total_amt  store_type  
transaction_id                                          
80712190438     -772.0  405.300  -4265.300           1  
29258453508    -1497.0  785.925  -8270.925           1  
51750724947     -791.0  166.110  -1748.110           2  
93274880719    -1363.0  429.345  -4518.345           1  
51750724947     -791.0  166.110  -1748.110           2

In [29]:
df['day'] = df['tran_date'].apply(lambda x : x.split('-')[0])
df['month'] = df['tran_date'].apply(lambda x : x.split('-')[1])
df['year'] = df['tran_date'].apply(lambda x : x.split('-')[2])
df

cust_id   tran_date  prod_subcat_code  prod_cat_code  qty  \
transaction_id                                                              
80712190438      270351  28-02-2014                 1              1   -5   
29258453508      270384  27-02-2014                 5              3   -5   
51750724947      273420  24-02-2014                 6              5   -2   
93274880719      271509  24-02-2014                11              6   -3   
51750724947      273420  23-02-2014                 6              5   -2   

                  rate      tax  total_amt  store_type day month  year  
transaction_id                                                          
80712190438     -772.0  405.300  -4265.300           1  28    02  2014  
29258453508    -1497.0  785.925  -8270.925           1  27    02  2014  
51750724947     -791.0  166.110  -1748.110           2  24    02  2014  
93274880719    -1363.0  429.345  -4518.345           1  24    02  2014  
51750724947     -791.0  166.110  -1748.110           2  23    02  2014

In [30]:
df['unit_price'] = df.apply(lambda row : row['total_amt']/row['qty'], axis = 1)
df

cust_id   tran_date  prod_subcat_code  prod_cat_code  qty  \
transaction_id                                                              
80712190438      270351  28-02-2014                 1              1   -5   
29258453508      270384  27-02-2014                 5              3   -5   
51750724947      273420  24-02-2014                 6              5   -2   
93274880719      271509  24-02-2014                11              6   -3   
51750724947      273420  23-02-2014                 6              5   -2   

                  rate      tax  total_amt  store_type day month  year  \
transaction_id                                                           
80712190438     -772.0  405.300  -4265.300           1  28    02  2014   
29258453508    -1497.0  785.925  -8270.925           1  27    02  2014   
51750724947     -791.0  166.110  -1748.110           2  24    02  2014   
93274880719    -1363.0  429.345  -4518.345           1  24    02  2014   
51750724947     -791.0  166.110  -1748.110           2  23    02  2014   

                unit_price  
transaction_id              
80712190438        853.060  
29258453508       1654.185  
51750724947        874.055  
93274880719       1506.115  
51750724947        874.055

In [39]:
date1 = df.tran_date.iloc[0]

'28-02-2014'

# Data Cleaning and Missing Values Management

In [ ]:
# Insert your code here
import pandas as pd

df = pd.read_csv('transactions.csv', sep=',', index_col='transaction_id')
df1 = df.copy()
df.head(10)

## Cleaning up a dataset

In [ ]:
1. Cleaning up a dataset

    In this part we will introduce the methods of the DataFrame class that are essential to clean a dataset. These methods can be grouped into three different categories :

            Duplicates management (duplicated and drop_duplicates methods)

            Modification of the elements of a DataFrame (replace, rename and astype methods)

            Operations on the values of a DataFrame (apply method and lambda functions)

Managing duplicates (duplicated and drop_duplicates methods)

    Duplicates are identical entries that appear more than once in a dataset.

    When we first discover a dataset it is very important to check up front that there are no duplicates. The presence of duplicates will generate errors in the computation of statistics or the plotting of graphs.

    Let df be the following DataFrame:
    	Age 	Gender 	Height
    Robert 	56 	M 	174
    Mark 	23 	M 	182
    Alina 	32 	F 	169
    Mark 	23 	M 	182

    The presence of duplicates is checked using the duplicated method of a DataFrame:

    # We identify the rows containing duplicates
    df.duplicated()

    >>> 0 False
    >>> 1 False
    >>> 2 False
    >>> 3 True

    This method returns Series object from pandas, which is equivalent to the column of a DataFrame. The Series object tells us for each row wether it is a duplicate.

    In this example, the result of the duplicated method informs us that the row with index 3 is a duplicate. Indeed, it is the exact copy of the row with index 1.

    Since the duplicated method returns an object of the Series class, we can apply the sum method to it in order to count the number of duplicates:

    # To calculate the sum of boolean values, we consider that True is worth 1 and False is worth 0.
    print(df.duplicated().sum())
    >>> 1

    The method of the DataFrame class used to remove duplicates is drop_duplicates. Its header is as follows:

    drop_duplicates(subset, keep, inplace)

            The subset parameter indicates the column(s) to consider in order to identify and remove duplicates. By default, subset = None namely we consider all the columns of the DataFrame.

            The keep parameter indicates which entry should be kept :

                    'first' : We keep the first occurrence.
                    'last': We keep the last occurrence.
                    False: We do not keep any occurrence.
                    By default, keep = 'first'.

            The inplace parameter (very common in the methods of the DataFrame class), specifies whether you modify directly the DataFrame (in this case inplace = True) or if the method returns a copy of the DataFrame (inplace = False). A method applied with the argument inplace = True is irreversible. By default, inplace = False.

    You have to be very careful when using the inplace parameter. A good practice is to forget this parameter and assign the DataFrame returned by the method to a new DataFrame.

    The keep parameter is the one that is most often specified. Indeed, a database can have duplicates created on different dates. We will then specify the value of the keep argument to keep only the most recent entries, for example.

    Let us go back to the df DataFrame :
    	Age 	Gender 	Height
    Robert 	56 	M 	174
    Mark 	23 	M 	182
    Alina 	32 	F 	169
    Mark 	23 	M 	182

    We illustrate df with the following illustration :

    We illustrate in the following examples the entries that are deleted by the drop_duplicates method depending on the value of the keep parameter:

    # We keep only the first occurrence of the duplicate
    df_first = df.drop_duplicates(keep = 'first')



    # We keep only the last occurrence of the duplicate
    df_last = df.drop_duplicates(keep = 'last')

    # We keep no duplicates
    df_false = df.drop_duplicates(keep = False)

    (a) How many duplicates are there in the transactions DataFrame ?



In [ ]:
# Counting the number of duplicates
duplicates = transactions.duplicated().sum()

print ("There are", duplicates, "duplicates in transactions.")

In [ ]:


    The transactions were recorded in anti-chronological order, i.e. the first rows contain the most recent transactions and the last rows the oldest transactions.

    (b) Eliminate duplicates from the database by keeping only the first occurrence, i.e. the most recent transaction.

    (c) Using the subset and keep parameters of the drop_duplicates method of transactions, display the most recent transaction for each category of prod_cat_code. To do this, you can remove all the duplicates from the prod_cat_code column by keeping only the first occurrence.



In [ ]:
transactions = transactions.drop_duplicates(keep = 'first')


transactions.drop_duplicates(subset = ['prod_cat_code'], keep = 'first')

### Modification of the elements of a DataFrame (replace, rename and astype methods)

In [ ]:
Modification of the elements of a DataFrame (replace, rename and astype methods)

    The replace method allows to replace one or more values ​​of a column of aDataFrame.

    Its header is as follows:

    replace(to_replace, value, ...)

            The to_replace parameter contains the value or the list of values to be replaced. It can be a list of integers, strings, booleans, etc.

            The value parameter contains the value or the list of the substitute values. It can also be a list of integers, strings, booleans, etc.



    In addition to modifying the elements of a DataFrame, it is possible to rename its columns.

    This is possible thanks to the rename method which takes as argument a dictionary whose keys are the old names and the values are the new names. You must also fill in the argument axis = 1 to specify that the names to rename are those of the columns.

    # Creation of the dictionary associating the old names with the new column names
    dictionary = {'old_name1': 'new_name1',
                  'old_name2': 'new_name2'}

    # We rename the variables using the rename method
    df = df.rename(dictionary, axis = 1)

    It is sometimes necessary to modify not only the name of a column but also its type.

    For example, it is possible that when importing a database, a variable is of type string when in fact it is a numerical variable. Whenever one of the entries in the column is incorrectly recognized, pandas will consider that this column is of type string.

    This is possible thanks to the astype method.

    The types that we will see most often are:

            str: Character string ('Hello').
            float: Floating point number (1.0, 1.14123).
            Int: Integer (1,1231)

    As for the rename method, astype can take as argument a dictionary whose keys are the names of the columns whose type should be modified and the values are the new types to assign. This is useful if you want to change the type of several columns at once.

    Most often, we will directly select the column whose type should be modified and overwrite it by applying the astype method to it.

    # Method 1: Creation of a dictionary then call to the astype method of the DataFrame
    dictionary = {'col_1': 'int',
                  'col_2': 'float'}
    df = df.astype(dictionary)

    # Method 2: Selection of the column and then calling the astype method of a Series
    df['col_1'] = df['col_1'].astype('int')

    These methods also have the inplace parameter to perform the operation directly on the DataFrame. To be used with great caution.

    If you make a mistake in the next exercise, you can re-import and redo the preprocessing by running the following cell.



In [ ]:
# Data import
transactions = pd.read_csv("transactions.csv", sep = ',', index_col = "transaction_id")

# Removal of duplicates
transactions = transactions.drop_duplicates(keep = 'first')

In [ ]:
    d) Import the numpy module under the namenp.

    (e) Replace the modalities ['e-Shop', 'TeleShop', 'MBR', 'Flagship store', np.nan] of the Store_type column by the modalities [1, 2, 3, 4, 0].

        The np.nan value is the one that encodes a missing value. We will replace this value with 0.

    (f) Convert the type of the columns Store_type and prod_subcat_code to type 'int'.

    (g) Rename the 'Store_type','Qty', 'Rate' and 'Tax' columns with 'store_type','qty', 'rate' and 'tax'.


In [ ]:
# Data import
transactions = pd.read_csv("transactions.csv", sep = ',', index_col = "transaction_id")

# Removal of duplicates
transactions = transactions.drop_duplicates(keep = 'first')

## Exercise

import numpy as np

# Replacement of values
transactions = transactions.replace(to_replace = ['e-Shop', 'TeleShop', 'MBR', 'Flagship store', np.nan],
                                    value = [1, 2, 3, 4, 0])

# Conversion of column types
new_types = {'Store_type'       : 'int',
             'prod_subcat_code' : 'int'}

transactions = transactions.astype(new_types)

# Renaming the columns
new_names = {'Store_type'   : 'store_type',
              'Qty'         : 'qty',
              'Rate'        : 'rate',
              'Tax'         : 'tax'}

transactions = transactions.rename(new_names, axis = 1)

# Display of the first rows of transactions
transactions.head()

### Operations on the values of a DataFrame (apply method and lambda functions)

In [ ]:
Operations on the values ​​of a DataFrame (apply method and lambda functions)

    It is often interesting to modify or aggregate the information of the columns of a DataFrame using an operation or a function.

    These operations can be any type of function which takes a column as argument. Thus, the numpy module is perfectly suited to perform operations on this type of object.

    The method used to perform an operation on a column is the apply method of a DataFrame whose header is:

    apply(func, axis, ...)

    where:

            func is the function to apply to the column.
            axis is the dimension on which the operation must be applied.

    Example: apply and np.sum

    For each column with numerical values, we want to calculate the sum of all rows. The sum function of numpy does this, so we can use it with the apply method.

    Since we are going to perform an operation on the rows, we must therefore specify the argument axis = 0 in the apply method.

    # Sum of the ROWS for each column of df
    df_lines = df.apply(np.sum, axis = 0)

    The result is the following:


    Now, for each row we want to compute the sum of all the columns.

    We are going to perform this operation on the columns, we must therefore specify the argument axis = 1 in the apply method.

    # Sum of columns for each ROW of df
    df_columns = df.apply(np.sum, axis = 1)

    The result is the following:

    These examples only illustrate the use of the apply method. To actually compute the sum of rows or columns, it is better to use the sum method of a DataFrame or a Series, which behaves in exactly the same way as the sum method of a numpy array.

The tran_date column of transactions contains the dates of the transactions in the format ('day-month-year') (ex: '28-02-2014'). The dates are of type string: it is not possible to perform statistics on this variable for the moment.

We would rather have 3 different columns for the day, month and year of each transaction. This would allow us, for example, to analyze and detect trends in transaction dates.

The date '28-02-2014' is a string. The day, month and year are separated by a hyphen '-'. The character string class has the split method to split a string on a specific character:

date = '28-02-2014 '

# Splitting the string on the '-' character
print(date.split('-'))
>>> ['28', '02', '2014']

This method returns a list containing the slices of the string on the specified character. Thus, to retrieve the day, all you have to do is select the first element of the split. To recover the month, we must take the second element and for the year the third.

    (h) Define a function get_day taking as argument a string and which returns the first element of its split by the character '-'.

    (i) Define the functions get_month and get_year which do the same with the second and third element of the split.

    (j) In 3 variables called days, months and years, store the result of the apply method on the tran_date column with the get_day, get_month and get_year functions. As these functions work element-wise, it is not necessary to specify the argument axis in the apply method.

    (k) Create the columns 'day', 'month' and 'year' in the transactions DataFrame and store the values of the days, months and years. Creating a new column is simply done by declaring it:

        # Create a new column 'day' with the values contained in days.
        transactions['day'] = days

    (l) Display the first 5 rows of transactions.



In [ ]:
# Definition of the functions to apply to the 'tran_date' column
def get_day(date):
    """
    Takes a date as a string argument.
    
    The date must have the format 'DD-MM-YYYY'.
    
    This function returns the day (DD).
    """
    
    # Splitting the string on the '-' character
    splits = date.split('-')
    
    # We return the first element of the breakdown (day)
    day = splits[0]
    return day

def get_month(date):
    return date.split('-')[1]

def get_year(date):
    return date.split('-')[2]
    
    
# Retrieving the day, month and year of each transaction
days = transactions['tran_date'].apply(get_day)
months = transactions['tran_date'].apply(get_month)
years = transactions['tran_date'].apply(get_year)

# Creation of new columns
transactions['day'] = days
transactions['month'] = months
transactions['year'] = years

# Displaying first rows of transactions
transactions.head()

In [ ]:


    The apply method is very powerful when combined with a lambda function.

    In Python, the keyword lambda is used to define an anonymous function: a function declared without name.

    A function lambda can take any number of arguments, but can only have one expression.

    Here is its syntax:

    lambda arguments: expression

    Lambda functions allow you to define functions with a very short syntax :

    # Example 1 
    x = lambda a: a + 2
    print(x(3))
    >>> 5

    # Example 2 
    x = lambda a, b : a * b
    print(x(2, 3))
    >>> 6

    # Example 3 
    x = lambda a, b, c : a - b + c
    print(x(1, 2, 3))
    >>> 2

    Although syntactically different, lambda functions behave in the same way as regular functions that are declared using the def keyword.

    The classic definition of a function is done with the def keyword:

    def increment(x):
        return x + 1

    It is also possible to define a function with the keyword lambda:

    increment = lambda x: x + 1

    The first method is very clean but the advantage of the second is that it can be defined on-the-fly directly within the apply method.

    Thus, the previous exercise can be done with a very compact syntax:

    transactions['day'] = transactions['tran_date'].apply(lambda date: date.split('-')[0])

    This kind of syntax is very practical and very often used for cleaning databases.

    The prod_subcat_code column of transactions depends on the prod_cat_code column because it identifies a subcategory of product. It would make more sense to have the category and subcategory of a product in the same variable.

    To do this, we will merge the values of these two columns:

            We will first convert the values of these two columns into strings using the astype method.

            Then, we will concatenate these strings to have a unique code representing both the category and sub-category. This can be done in the following way:

        string1 = "I think"
        string2 = "therefore I am."

        # Concatenation of the two strings by separating them with a space
        print (string1 + " " + string2)
        >>> I think therefore I am.

To apply a lambda function to an entire row, you must specify the argument axis = 1 in the apply method. In the function itself, the columns of the row can be accessed as on a DataFrame:

# Computation of the unit price of a product
transactions.apply(lambda row: row['total_ amt']/row['qty'], axis = 1)

    (m) Using a lambda function applied to transactions, create a column 'prod_cat' in transactions containing the concatenation of the values ofprod_cat_code and prod_subcat_code separated by a hyphen '-'. Remember to convert the values to strings.

    Displaying this column should yield:

    transaction_id
    80712190438     1-1
    29258453508     3-5
    51750724947     5-6
    93274880719     6-11
    51750724947     5-6
                   ...
    94340757522     5-12
    89780862956     1-4
    85115299378     6-2
    72870271171     5-11
    77960931771     5-11



In [ ]:
# Insert your code here
transactions['prod_cat'] = transactions.apply(lambda row : str(row['prod_cat_code']) + '-' + str(row['prod_subcat_code']),
                                    axis=1)
transactions.head()

# original cozum
transactions['prod_cat'] = transactions.astype('str').apply(lambda row: row['prod_cat_code']+'-'+row['prod_subcat_code'],
                                                            axis = 1)

print(transactions['prod_cat'])

## Dealing with missing values

In [ ]:
2. Dealing with missing values

    A missing value is either:

            An unspecified value.
            A value that does not exist. In general, they result from mathematical calculations having no solution (a division by zero for example).

    A missing value appears under the name NaN ("Not a Number") in a DataFrame.

    In this part, we will see several methods to:

            Detect missing values (isna and any methods)
            Replace these values (fillna method)
            Delete missing values (dropna method)

    In one of the previous exercises, we used the replace method of transactions to replace missing values with 0. This approach is not rigorous and should not be done in practice.

    For this reason, we are going to re-import the raw version of transactions to undo the steps we did in the previous exercises.

    (a) Run the cell below to re-import transactions, remove duplicates and rename its columns.



In [ ]:
# Data import
transactions = pd.read_csv("transactions.csv", sep = ',', index_col = "transaction_id")

# Duplicates removal
transactions = transactions.drop_duplicates(keep = 'first')

# Renaming the columns
new_names = {'Store_type'  : 'store_type',
              'Qty'        : 'qty',
              'Rate'       : 'rate',
              'Tax'        : 'tax'}

transactions = transactions.rename(new_names, axis = 1)

transactions.head()

### Detecting missing values (isna and any methods)

In [ ]:
Detecting missing values (isna and any methods)

    The isna method of a DataFrame detects its missing values. This method does not take any arguments.

    This method returns the same DataFrame whose values are:

            True if the original table cell is a missing value (np.nan)
            False otherwise.



    Since the isna method returns a DataFrame, we can use it with other methods of the DataFrame class to get more precise information:

            The any method - thanks to its axis argument - allows to determine which columns (axis = 0) or which rows (axis = 1) contain at least one missing value.

            The sum method counts the number of missing values per column or row (by specifying the axis argument). It is possible to use other statistical methods like mean,max, argmax, etc.

    Here are many examples of using the any and sum methods with isna:

    We use the DataFrame df from the previous illustrations:
    	Name 	Country 	Age
    0 	NaN 	Australia 	NaN
    1 	Duchamp 	France 	25
    2 	Hana 	Japan 	54

    The df.isna() instruction returns:
    	Name 	Country 	Age
    0 	True 	False 	True
    1 	False 	False 	False
    2 	False 	False 	False

    # COLUMNS containing at least one missing value are detected
    df.isna().any(axis = 0)

    >>> Name     True
    >>> Country  False
    >>> Age      True

    # ROWS containing at least one missing value are detected
    df.isna().any (axis = 1)

    >>> 0    True
    >>> 1    False
    >>> 2    False

    # Using conditional indexing to display entries
    # containing missing values
    df[df.isna().any(axis = 1)]

    which returns the DataFrame:
    	Name 	Country 	Age
    0 	NaN 	Australia 	NaN

    # We count the number of missing values for each COLUMN
    df.isnull().sum(axis = 0)

    >>> Name     1
    >>> Country  0
    >>> Age      1

    # Count the number of missing values for each ROW
    df.isnull().sum(axis = 1)

    >>> 0   2
    >>> 1   0
    >>> 2   0

    The methods isnaand isnullhave exactly the same behavior.

    (b) How many columns of the transactions DataFrame contain missing values?

    (c) How many of transactions' entries contain missing values? You can use the any method along with the sum method.

    (d) Which column of transactions contains the most of missing values? You can use the idxmax method that returns the index of first occurrence of maximum over the requested axis.

    (e) Show transaction entries that contain at least one missing value in the 'rate', 'tax' and 'total_amt' columns. What do you notice?



In [ ]:
# Which columns contain NaNs
columns_na = transactions.isna().any(axis = 0)

print(columns_na.sum(), "columns of transactions contain NaNs. \n")

# Which rows contain NaNs
rows_na = transactions.isna().any(axis = 1)

print(rows_na.sum(), "rows of transactions contain NaNs. \n")

# Number of NaNs per column
columns_nbna = transactions.isna().sum(axis = 0)

print ("The column containing the most NaNs is:",  columns_nbna.idxmax())

# Display the first 10 entries containing at least one NaN in 'rate', 'tax' or 'total_amt'
transactions[transactions[['rate', 'tax', 'total_amt']].isna().any(axis = 1)].head(10)

# The three variables are still missing together.

### Replacing missing values (fillna method)

In [ ]:
Replacing missing values (fillna method)

    The fillna method allows you to replace the missing values of a DataFrame by values you want.

    # We replace all the NaNs of the DataFrame by zeros
    df.fillna(0)

    # We replace the NaNs of each numerical column by the average on this column
    df.fillna(df.mean()) # df.mean() can be replaced by any statistical method.

    It is common to replace missing values of a column containing numerical values with statistics like:

            The mean: .mean
            The median: .median
            The minimum / maximum: .min / .max.

    For categorical type columns, replace the missing values with:

            The mode, i.e. the most frequent modality: .mode.
            A constant or arbitrary category: 0,-1.

    To avoid making replacement errors, it is very important to select the correct columns before using the fillna method.

If you make mistakes in the following exercise, you can re-import transactions using the following cell:


In [ ]:
# Data import
transactions = pd.read_csv("transactions.csv", sep = ',', index_col = "transaction_id")

# Removal of duplicates
transactions = transactions.drop_duplicates(keep = 'first')

# Renaming the columns
new_names = {'Store_type' : 'store_type',
              'Qty'       : 'qty',
              'Rate'      : 'rate',
              'Tax'       : 'tax'}

transactions = transactions.rename(new_names, axis = 1)

In [ ]:

    (f) Replace the missing values in prod_subcat_code column of transactions with -1.

    (g) Determine the most frequent modality (the mode) of the store_type column of transactions.

    (h) Replace the missing values of the store_type column by this modality. The value of this modality is accessed at index 0 of the Series returned by mode.

    (i) Check that the prod_subcat_code and store_type columns of transactions no longer contain missing values.



In [ ]:
# Replacing the NaNs of 'prod_subcat_code' by -1
transactions['prod_subcat_code'] = transactions['prod_subcat_code'].fillna(-1)

# Determining the mode of 'store_type'
store_type_mode = transactions['store_type'].mode()
print ("The most frequent mode of 'store_type' is:", store_type_mode[0])

# Replacing the NaNs of 'store_type' by its mode
transactions['store_type'] = transactions['store_type'].fillna(transactions['store_type'].mode()[0])

# Checking that these two columns no longer contain NANs
transactions[['prod_subcat_code', 'store_type']].isna().sum()

### Removing missing values (dropna method)

In [ ]:
Removing missing values (dropna method)

    The dropna method allows you to remove rows or columns containing missing values.

    The header of the method is as follows:

    dropna(axis, how, subset, ..)

            The axis parameter specifies whether to delete rows or columns (0 for rows, 1 for columns).

            The how parameter lets you specify how the rows (or columns) are deleted:

                    how = 'any': We delete the row (or column) if it contains at least one missing value.
                    how = 'all': We delete the row (or column) if it contains only missing values.

            The subset parameter is used to specify the columns/rows on which the search for missing values is carried out.

    Example:

    # We delete all the rows containing at least one missing value
    df = df.dropna(axis = 0, how = 'any')

    # We delete the empty columns
    df = df.dropna(axis = 1, how = 'all')

    # We remove the rows with missing values in the 3 columns 'col2', 'col3' and 'col4'
    df.dropna(axis = 0, how = 'all', subset = ['col2', 'col3', 'col4'])

    As with the other methods of replacing values of a DataFrame, the inplace argument can be used with great care to perform the modification directly without reassignment.

Transaction data for which the transaction amount is not provided is of no interest to us. For this reason:

    (j) Delete the transaction entries for which the rate, tax and total_amt columns are all empty.

    (k) Check that the columns of transactions no longer contain missing values.



In [ ]:
transactions = transactions.dropna(axis = 0, how = 'all', subset = ['rate', 'tax', 'total_amt'])

transactions.isna().sum(axis = 0)

## Conclusion and recap

In [ ]:
Conclusion and recap

    In this chapter we have seen the essential methods of the pandas module in order to clean up a dataset and manage missing values ​​(NaN).

    This step of preparing a dataset is always the first step of a data project.

    Regarding data cleaning, we have learned how to:

            Identify and delete duplicates of a DataFrame using the duplicated and drop_duplicates methods.

            Modify the elements of a DataFrame and their type using the replace, rename and astype methods.

            Apply a function to a DataFrame with the apply method and the lambda clause.

    Regarding the management of missing values, we have learned to:

            Detect them using the isna method followed by the any and sum methods.

            Replace them using the fillna method and the statistical methods.

            Delete them using the dropna method.

    In the following notebook, you will see other manipulations of DataFrames for a more advanced exploration of data.



## Apply, Lamda, Func Ornekleri (EXTRA)

In [85]:
import numpy as np
import pandas as pd 

df = pd.read_csv('Data\transactions.csv')

def get_day(string):
    day = string.split('-')[0]
    return day

def get_day_column(df, column):
    lst1 = [i.split('-')[0] for i in column]
    df['day2'] = np.array(lst1)
    return df['day2']
        
date1 = df.tran_date.iloc[0]
date1

print(get_day(date1))
# 1. Yol
df['day'] = df.tran_date.apply(get_day)

# 2. Yol
get_day_column(df, df.tran_date.values)

# 3. Yol 
df['day'] = df['tran_date'].apply(lambda x : x.split('-')[0])

# 4.yol
df[['day', 'month', 'year']] = df['tran_date'].str.split(expand=True)


28


transaction_id
80712190438    28
29258453508    27
51750724947    24
93274880719    24
51750724947    23
Name: day2, dtype: object

In [70]:
df.tran_date.values

array(['28-02-2014', '27-02-2014', '24-02-2014', '24-02-2014',
       '23-02-2014'], dtype=object)

In [83]:
df

cust_id   tran_date  prod_subcat_code  prod_cat_code  qty  \
transaction_id                                                              
80712190438      270351  28-02-2014                 1              1   -5   
29258453508      270384  27-02-2014                 5              3   -5   
51750724947      273420  24-02-2014                 6              5   -2   
93274880719      271509  24-02-2014                11              6   -3   
51750724947      273420  23-02-2014                 6              5   -2   

                  rate      tax  total_amt  store_type day month  year  \
transaction_id                                                           
80712190438     -772.0  405.300  -4265.300           1  28    02  2014   
29258453508    -1497.0  785.925  -8270.925           1  27    02  2014   
51750724947     -791.0  166.110  -1748.110           2  24    02  2014   
93274880719    -1363.0  429.345  -4518.345           1  24    02  2014   
51750724947     -791.0  166.110  -1748.110           2  23    02  2014   

                unit_price day2  
transaction_id                   
80712190438        853.060   28  
29258453508       1654.185   27  
51750724947        874.055   24  
93274880719       1506.115   24  
51750724947        874.055   23

In [38]:
import pandas as pd
#df_names = pd.read_clipboard()
df_names

Name         Car
0      Lila      Twingo
1     Tiago        Clio
2  Berenice   C4 Cactus
3    Joseph      Twingo
4     Kader       Swift
5      Romy      Scenic

In [37]:
#df_cars = pd.read_clipboard()
df_cars

Car  Price
0     Twingo   11000
1      Swift   14500
2  C4 Cactus   23000
3       Clio   16000
4      Prius   30000

# Extra Filtering

In [ ]:
listem = ['y', 'n'] # İstenen veri listesi
count = 0 # Sayacı sıfırla
for col in df.iloc[:, 1:].columns: # Sütunları döngüye sok
    for val in df[col].values: # Her bir hücreyi döngüye sok
        if val not in listem: # Eğer hücrenin değeri listede yoksa
            count += 1 # Sayacı arttır
count


count = sum([sum([1 for val in col if val not in ['y', 'n']]) for col in df.values])
count


check = all(all(val in ['y', 'n'] for val in col) for col in df.values)
check


df.apply(lambda x : x.isin(['y', 'n']) == False) # bitmedi daha???

# Data Processing

In [ ]:


    Data processing can be reduced to the use of 4 essential operations: filtering, merging, ordering and grouping.

    If the DataFrame class has prevailed in the domain of data manipulation, it is because it is often sufficient to repeat or combine these four operations.

    In this exercise, you will learn how to use these 4 operations of data processing.

    Before starting this notebook, run the following cell in order to retrieve the work done in the previous exercises.


In [ ]:
### Imports ###

import pandas as pd

# Data import
transactions = pd.read_csv("transactions.csv", sep =',', index_col = "transaction_id")

# Removal of duplicates
transactions = transactions.drop_duplicates(keep = 'first')

# Changing the names of the columns
new_names =  {'Store_type' : 'store_type',
              'Qty'        : 'qty',
              'Rate'       : 'rate',
              'Tax'        : 'tax'}

transactions = transactions.rename(new_names, axis = 1)

### Handling NAs ###

#  We replace the NAs in 'prod_subcat_code' by -1
transactions['prod_subcat_code'] = transactions['prod_subcat_code'].fillna(-1).astype(int)

# We compute the mode of 'store_type'
store_type_mode = transactions['store_type'].mode()

# We replace the NAs of 'store_type' by its mode
transactions['store_type'] = transactions['store_type'].fillna(transactions['store_type'].mode()[0])

# Removal of rows where 'rate', 'tax' and 'total_amt' are all NAs
transactions = transactions.dropna(axis = 0, how = 'all', subset = ['rate', 'tax', 'total_amt'])

## 1. Filtering a DataFrame with binary operators.

    Filtering consists in selecting a subset of rows of a DataFrame which meet a condition. Filtering corresponds to what was called conditional indexing until now, but the term "filtering" is the one that is used most in database management.

    We cannot use the logical operators and and or to filter on multiple conditions. Indeed, these operators create ambiguities that pandas is unable to handle for filtering.

    The operators suitable for filtering on several conditions are the binary operators:

            The 'and' operator: &.

            The 'or' operator: |.

            The 'not' operator: -.

    These operators are similar to logical operators but their evaluation methods are not the same.

The 'and' operator: &.

    The & operator is used to filter a DataFrame on several conditions which must be verified simultaneously.

    Example:

    Consider the following DataFrame`` df that contains information on apartments in Paris:
    	district 	year 	surface
    0 	'Champs-Elysées' 	1979 	70
    1 	'Europe' 	1850 	110
    2 	'Père-Lachaise' 	1935 	55
    3 	'Bercy' 	1991 	30

    If we want to find an apartment dating from 1979 and with a surface area greater than 60 squared meters, we can filter the lines of df with the following code:


    # Filtering of the DataFrame on the 2 previous conditions
    print(df[(df['year'] == 1979) & (df['surface']> 60)])

    >>>         district  year  surface
    >>> 0  Champs-Elysées  1979       70

    The conditions must be written between parentheses to eliminate any ambiguity on the order of evaluation of the conditions. Indeed, if the conditions are not properly separated, we will get the following error:

    print(df[df['year'] == 1979 & df['surface']> 60])

    >>> ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

The 'or' operator: |.

    The operator | is used to filter a DataFrame on several conditions of which one at least must be verified.

    Example:

    Consider the same DataFrame df:
    	district 	year 	surface
    0 	'Champs-Elysées' 	1979 	70
    1 	'Europe' 	1850 	110
    2 	'Père-Lachaise' 	1935 	55
    3 	'Bercy' 	1991 	30

    If we want to find an apartment that dates after 1900 or is located in the Père-Lachaise district, we can filter the lines of df with the following code:


    # Filtering of the DataFrame on the 2 previous conditions
    print(df[(df['year']> 1900) | (df['district'] == 'Père-Lachaise')])

    >>>          district  year  surface
    >>> 0  Champs-Elysées  1979       70
    >>> 2  Père-Lachaise   1935       55
    >>> 3  Bercy           1991       30

The 'not' operator: -.

    The operator - is used to filter aDataFrame on a condition which must not be true, i.e. whose negation must be verified.

    Example:

    Consider the same DataFrame`` df:
    	district 	year 	surface
    0 	'Champs-Elysées' 	1979 	70
    1 	'Europe' 	1850 	110
    2 	'Père-Lachaise' 	1935 	55
    3 	'Bercy' 	1991 	30

    If we want to retrieve apartments not located in the Bercy district then we filter df as follows:

    # Filtering of the DataFrame on the the negation of a condition
    print(df[-(df['district'] == 'Bercy')])

    >>> district year surface
    >>> 0 Champs-Elysées 1979 70
    >>> 1 Europe 1850 110
    >>> 2 Père-Lachaise 1935 55

    (a) Display the first 5 lines of the transactions DataFrame.

    (b) From transactions, create a DataFrame named e_shop containing only the transactions carried out in stores of type 'e-Shop' with a total amount greater than than 5000 ('store_type' and 'total_ amt' columns).

    (c) Similarly, create a DataFrame named teleshop which contains the transactions made in stores of type 'TeleShop' with a total amount of more than 5000.

    (d) Which of the two types of store has the most transactions over € 5,000?



In [ ]:
# Insert your code here
df = transactions.copy()

print('solution a', 50 * '-')

(df.head())

print('solution b', 50 * '-')

e_shop = df[(df['store_type'] == 'e-Shop') & (df['total_amt'] > 5000)]
e_shop

print('solution c', 50 * '-')

tele_shop = df.loc[(df['store_type'] == 'TeleShop') & (df['total_amt'] > 5000)]
tele_shop

print('solution d', 50 * '-')

print('e_shop shape :', e_shop.shape[0], 'tele_shop shape :', tele_shop.shape[0])


#Original Cozum

# Creation of e_shop et teleshop
e_shop = transactions[(transactions['store_type'] == 'e-Shop') & (transactions['total_amt'] > 5000)]


teleshop = transactions[(transactions['store_type'] == 'TeleShop') & (transactions['total_amt'] > 5000)]


# We count the number of rows of each DataFrame. Other solutions are possible.
print('Number of transactions over 5000€ for e-shop :', len(e_shop))
print('Number of transactions over 5000€ for teleshop :', len(teleshop))

### output 
Number of transactions over 5000€ for e-shop : 1185

Number of transactions over 5000€ for teleshop : 532



    (e) Import into two DataFrames named respectively customer and prod_cat_info the data contained in the files 'customer.csv' and 'prod_cat_info.csv'.

    (f) The Gender and city_code columns of customer contain two missing values each. Replace them with their mode using the fillna and mode methods.



In [ ]:
# Insert your code here
print('solution e', 50 * '-')
customer = pd.read_csv('customer.csv')
customer.head()
prod_cat_info = pd.read_csv('prod_cat_info.csv')
prod_cat_info.head()

print('solution f', 50 * '-')
mode_gender = customer['Gender'].mode()[0]
customer['Gender'].fillna(mode_gender, inplace=True)

mode_city_code = customer['city_code'].mode()[0]
customer['city_code'].fillna(mode_city_code, inplace=True)

customer.isna().sum()

In [ ]:
# Output
prod_cat_code 	prod_cat 	prod_sub_cat_code 	prod_subcat
0 	1 	Clothing 	4 	Mens
1 	1 	Clothing 	1 	Women
2 	1 	Clothing 	3 	Kids
3 	2 	Footwear 	1 	Mens
4 	2 	Footwear 	3 	Women

customer_Id    0
DOB            0
Gender         0
city_code      0
dtype: int64

In [ ]:
# Original Cozum
customer = pd.read_csv('customer.csv')
prod_cat_info = pd.read_csv('prod_cat_info.csv')

customer['Gender'] = customer['Gender'].fillna(customer['Gender'].mode()[0])
customer['city_code'] = customer['city_code'].fillna(customer['city_code'].mode()[0])

## 2. Joining Dataframes: concat function and merge method.
Concatenation of DataFrames with concat

    The concat function of thepandas module allows you to concatenate several DataFrames, i.e. juxtapose them horizontally or vertically.

    The header of this function is as follows:

    pandas.concat (objs, axis ..)

            The objs parameter contains the list of DataFrames to concatenate.
            The axis parameter specifies whether to concatenate vertically (axis = 0) or horizontally (axis = 1).


<img src="Photos\pd_concat_en.png" width="800" height="400">


    When the number of rows or columns of the DataFrames does not match, the concat function fills the empty cells with NaN, as shown in the illustration below.


<img src="Photos\pd_concat_none_en.png" width="800" height="400">

    (a) Split the columns of the transactions DataFrame in half with half of the columns in a DataFrame named part_1 and the second half in a DataFrame named part_2.

    (b) Reconstitute transactions in a DataFrame named union by concatenating part_1 and part_2 horizontally.

    (c) What happens if we concatenate part_1 and part_2 by filling in the argument axis = 0?



In [ ]:
# Insert your code here 

print('solution a', 50 * '-')
column_part1 = df.columns[:5]
part1 = df[column_part1]

column_part2  = df.columns[5:]
part2 = df[column_part2]

part1

print('solution b', 50 * '-')

union  = pd.concat([part1,part2], axis=1)
union

print('solution c', 50 * '-')

union_axis0 = pd.concat([part1,part2], axis=0)
union_axis0

# Original Cozum
# Splitting of the transactions DataFrame
part_1 = transactions[transactions.columns[:4]]
part_2 = transactions[transactions.columns[4:]]

# Reconstitution of the transactions DataFrame by concatenation
union = pd.concat([part_1,part_2], axis = 1)

# If we were to concatenate by filling in the argument "axis = 0", we would obtain a DataFrame where half of
# the valuers are NAs
#
# This is due to the fact that the argument 'axis = 0' forces the pd.concat function to create new ROWS
# in part_1 but it cannot fill them correctly since part_1 and part_2 have no columns in common.

## Merging DataFrames with the merge method

    Two DataFrames can be merged if they have a column in common. This is done thanks to the merge method of the DataFrame class whose header is as follows:

    merge(right, on, how, ...)

            The right parameter is the DataFrame to merge with the one calling the method.

            The on parameter is the name of the columns of the DataFrame which will be used as reference for the merge. They must be common to both DataFrames

            The how parameter allows you to choose the type of join to perform for merging the DataFrames. The values for this parameter are based on SQL syntax joins.


    The how parameter can take 4 values ('inner', 'outer', 'left', 'right') that we will illustrate on the two DataFrames named Persons and Vehicles below:
    
    Name 	Car
    Lila 	Twingo
    Tiago 	Clio
    Berenice 	C4 Cactus
    Joseph 	Twingo
    Kader 	Swift
    Romy 	Scenic
    
    Car 	Price
    Twingo 	11000
    Swift 	14500
    C4 Cactus 	23000
    Clio 	16000
    Prius 	30000

            'inner': The inner join returns the rows whose values in the common columns are present in the two DataFrames. This type of join is often not recommended because it can lead to the loss of many entries. However, the inner join does not produces NAs.

        The result of the inner join Persons.merge(right = Vehicles, on = 'Car', how = 'inner') is shown below:

<img src="Photos\pd_join_inner_en.png" width="700" height="700">


            'outer': The outer join Persons the two DataFrames in their entirety. No row will be deleted. This method can generate a lot of NAs.

        The result of the outer join Persons.merge(right = Vehicles, on = 'Car', how = 'outer') is shown below:

<img src="Photos\pd_join_outer_en.png" width="700" height="700">


            'left': The left join returns all the rows of the DataFrame on the left (i.e. the one calling the method), and completes them with the rows of the second DataFrame which coincide according to the values of the common column. This is the default value for the how parameter.

        The result of the left join Persons.merge(right = Vehicles, on = 'Car', how = 'left') is shown below:

<img src="Photos\pd_join_left_en.png" width="700" height="700">


            'right': The right join returns all the rows of the DataFrame on the right, and complete them with the rows of the left DataFrame which coincide according to the values of the common column.

        The result of the right join Persons.merge(right = Vehicles, on = 'Car', how = 'right') is shown below:
<img src="Photos\pd_join_right_en.png" width="700" height="700">


    Doing a left join, right join, or outer join followed by a dropna(how = 'any') is equivalent to an inner join.

    The customer DataFrame contains information about customers in the 'cust_id' column of transactions.

    The 'customer_Id' column of the customer DataFrame will be used to make the join between transactions and customer. This will enrich the transactions DataFrame with additional information.

    (d) Using the rename method and a dictionary, rename the 'customer_Id' column of the customer DataFrame to 'cust_id'.

    (e) Using the merge method, perform the left join between the DataFrames transactions and customer on the 'cust_id' column. Name the created DataFramefusion.

    (f) Did the merging produce NAs?

    (g) Display the first lines of fusion. What are the new columns?



In [ ]:
# Insert your code here

# d)
customer = customer.rename({'customer_Id' : 'cust_id'} , axis= 'columns')
customer

# e)
fusion = transactions.merge(customer, on='cust_id', how='left')
fusion

#f)
fusion.isna().sum()

# g)
fusion.head()

#original cozum
# We rename the 'customer_Id' column to 'cust_id' for merging
customer = customer.rename(columns = {'customer_Id': 'cust_id'})

# Left join between transactions and customer on the 'cust_id' column
fusion = transactions.merge(right = customer, on = 'cust_id', how = 'left')

# The merging did not produce NAs
fusion.isna().sum()

# The columns DOB, Gender, city_code have been added to transactions
fusion.head()


In [ ]:
cust_id 	tran_date 	prod_subcat_code 	prod_cat_code 	qty 	rate 	tax 	        total_amt 	store_type 	DOB      	Gender 	city_code
0 	270351 	28-02-2014 	1                       	1 	-5 	-772.0 	405.300 	-4265.300 	e-Shop 	        26-09-1981 	M 	5.0
1 	270384 	27-02-2014 	5                              	3 	-5 	-1497.0 785.925 	-8270.925 	e-Shop 	        11-05-1973 	F 	8.0
2 	273420 	24-02-2014 	6 	                        5 	-2 	-791.0 	166.110 	-1748.110 	TeleShop 	27-07-1992 	M 	8.0
3 	271509 	24-02-2014 	11 	                        6 	-3 	-1363.0 429.345 	-4518.345 	e-Shop 	        08-06-1981 	M 	3.0
4 	273420 	23-02-2014 	6 	                        5 	-2 	-791.0 	166.110 	-1748.110 	TeleShop 	27-07-1992 	M 	

In [ ]:


    The merging went well and produced no NaNs. However, the index of the DataFrame is no longer the column transaction_id' and has been reset with the default index (0,1, 2 , ...).

    It is possible to re-define the index of a DataFrame using the set_index method.

    This method can take as argument:

            The name of a column to use as indexing.
            A Numpy array or pandas Series with the same number of rows as the DataFrame calling the method.

    Example:

    Let df be the following DataFrame:
    	Name 	Car
    0 	Lila 	Twingo
    1 	Tiago 	Clio
    2 	Berenice 	C4 Cactus
    3 	Joseph 	Twingo
    4 	Kader 	Swift
    5 	Romy 	Scenic

    We can set the column 'Name' as being the new index:

    df = df.set_index('Name')

    This will produce the following DataFrame:



    Name 	Car
    Lila 	Twingo
    Tiago 	Clio
    Berenice 	C4 Cactus
    Joseph 	Twingo
    Kader 	Swift
    Romy 	Scenic

    We can also define the index from a Numpy array, from a Series, etc:

    # New index to use
    new_index = ['10000' + str(i) for i in range(6)]
    print(new_index)
    >>> ['100000', '100001', '100002', '100003', '100004', '100005']

    # Using an array or a Series is equivalent
    index_array = np.array(new_index)
    index_series = pd.Series(new_index)


    df = df.set_index(index_array)
    df = df.set_index(index_series)

    This will produce the following DataFrame:
    	Name 	Car
    100000 	Lila 	Twingo
    100001 	Tiago 	Clio
    100002 	Berenice 	C4 Cactus
    100003 	Joseph 	Twingo
    100004 	Kader 	Swift
    100005 	Romy 	Scenic

    To return to the default numeric indexing, we use the reset_index method of the DataFrame:

    df = df.reset_index()

    The indexing column that was used is not deleted. A new column will be created containing the old index:
    	index 	Name 	Car
    0 	100000 	Lila 	Twingo
    1 	100001 	Tiago 	Clio
    2 	100002 	Berenice 	C4 Cactus
    3 	100003 	Joseph 	Twingo
    4 	100004 	Kader 	Swift
    5 	100005 	Romy 	Scenic

Merging transactions andcustomer removed the index of transactions.

The index of a DataFrame can be retrieved using its .index attribute.

    (h) Take the index from transactions and use it to index fusion.



In [ ]:
# Insert your code here
fusio = fusion.set_index(transactions.index)
fusion

#Original; Cozum
# We retrieve the index of transactions
new_index = transactions.index

# We set the new index of fusion
fusion = fusion.set_index(new_index)
fusion.head()

## 3. Sort and order the values of a DataFrame: sort_values and sort_index methods.

    The sort_values method allows you to sort the rows of a DataFrame according to the values of one or more columns.

    The header of this method is as follows:

    sort_values(by, ascending, ...)

            The by parameter allows you to specify on which column(s) the sort is performed.

            The ascending parameter is a boolean value (True or False) determining whether the sorting order is ascending or descending. By default this parameter is set to True.

    Example:

    Consider the DataFrame df describing students:
    Name 	Grade 	Bonus points
    'Amelie' 	A 	1
    'Marin' 	F 	1
    'Pierre' 	A 	2
    'Zoe' 	C 	1

    First of all, we will sort the rows on a single column, for example the column 'Bonus Points':

    # We sort the DataFrame df on the column 'Bonus Points'
    df_sorted = df.sort_values(by ='Bonus Points', ascending = True)

    We obtain the following result:
<img src="Photos\pd_sort_values_en.png" width="700" height="500">

    The rows of the DataFrame df_sorted are therefore sorted in ascending order of the 'Bonus points' column. However, if we look at the column 'Grade', we see that it is not sorted alphabetically for the common values of 'Bonus Points'.

    This can be remedied by also sorting by the 'Grade' column:

    # We first sort the DataFrame df by the column 'Bonus Points' then in case of equality, by the column 'Grade'.
    df_sorted = df.sort_values(by = ['Bonus Points', 'Grade'], ascending = True)


    We obtain the following result:

<img src="Photos\pd_sort_values_2_en.png" width="700" height="500">



    The sort_index method allows you to sort a DataFrame according to its index.

    If the index is the default one (numerical), this method is not particularly interesting.

    However, it can often be combined with the set_index method of a DataFrame that we have just seen.

    Example:

    # We define the column 'Grade' as the index of df
    df = df.set_index('Grade')

    # We sort the DataFrame df according to its index
    df = df.sort_index()

    This produces the following DataFrame:



    Grade 	Name 	Bonus points
    A 	'Amelie' 	1
    A 	'Peter' 	2
    C 	'Zoe' 	1
    F 	'Sailor' 	1

    Consider the two following DataFrames containing boat rental data.

    Below are the boats DataFrame:
    	boat_name 	color 	reservation_number 	n_reservations
    0 	Julia 	blue 	2 	34
    1 	Siren 	green 	3 	10
    2 	Sea Sons 	red 	6 	20
    3 	Hercules 	blue 	1 	41
    4 	Cesar 	yellow 	4 	12
    5 	Minerva 	green 	5 	16

    And the clients DataFrame:
    	client_id 	client_name 	reservation_id
    0 	91 	Marie 	1
    1 	154 	Anna 	2
    2 	124 	Yann 	3
    3 	320 	Lea 	7
    4 	87 	Marc 	9
    5 	22 	Yassine 	10

    (a) Run the following cell to instantiate these DataFrames.



In [2]:
# Definition of the data dictionnaries
import pandas as pd
data_boats = {'boat_name'         : ['Julia', 'Siren', 'Sea Sons', 'Hercules', 'Cesar', 'Minerva'], 
              'color'             : ['blue', 'green', 'red', 'blue', 'yellow', 'green'],
              'reservation_number': [2, 3, 6, 1, 4, 5],
              'n_reservations'    : [34, 10, 20, 41, 12, 16]}

data_clients = {'client_id'     : [91, 154, 124, 320, 87, 22], 
                'client_name'   : ['Marie', 'Anna', 'Yann', 'Lea', 'Marc', 'Yassine'],
                'reservation_id': [1, 2, 3, 7, 9, 5]}

# Instantiation of the DataFrames
boats = pd.DataFrame(data_boats)
clients = pd.DataFrame(data_clients)




    We want to easily determine which customer has reserved the boats of the boats DataFrame. To do this, we can simply merge the DataFrames.

    (b) Rename the 'reservation_number' column from boats to 'reservation_id' using the rename method.

    (c) In a DataFrame named boats_clients, perform the left join between boats (left) and clients (right).

    (d) Set the column 'boat_name' as index of the boats_clients DataFrame.

    (e) Using the loc method, find who reserved the boats 'Julia' and 'Siren'.

    (f) Using the isna method applied to the client_name column, determine the boats that have not been reserved.

    (g) The number of times a boat has been reserved so far is indicated by the column 'n_reservations'. Using the sort_values method, determine the name of the customer who reserved the blue boat with the most reservations to date.



In [3]:
# Insert your code here
#b)
boats.rename(columns = {'reservation_number' :'reservation_id' }, inplace=True)

#c)
boats_clients = boats.merge(clients, how='left', on='reservation_id')
boats_clients
#d)
boats_clients = boats_clients.set_index('boat_name')
boats_clients
#e)
#boats_clients.loc[boats_clients.index.isin(['Julia', 'Siren'])]
boats_clients.loc[['Julia', 'Siren']]

#f)
#boats_clients.client_name.fillna('not_reserved', inplace=True)
boats_clients[boats_clients.client_name.isna()].index

#g)
boats_clients.sort_values('n_reservations', ascending=False)

color  reservation_id  n_reservations  client_id client_name
boat_name                                                               
Hercules     blue               1              41       91.0       Marie
Julia        blue               2              34      154.0        Anna
Sea Sons      red               6              20        NaN         NaN
Minerva     green               5              16       22.0     Yassine
Cesar      yellow               4              12        NaN         NaN
Siren       green               3              10      124.0        Yann

In [4]:
# Original Cozum
# We rename the column 'number_reservation'
boats = boats.rename(columns = {'reservation_number': 'reservation_id'})

# We perform the left join between boats and clients
boats_clients = boats.merge(clients, on = 'reservation_id', how = 'left')

# We set the column 'boat_name' as the index of boats_clients
boats_clients = boats_clients.set_index("boat_name")

# Who reserved 'Julia' and 'Siren'?
print("The client who reserved 'Julia' is:", boats_clients.loc['Julia', 'client_name'])
print("The client who reserved 'Siren' is:", boats_clients.loc['Siren', 'client_name'])
print("\n")

# Which boats have not been reserved?
boats_not_reserved = boats_clients[boats_clients['client_name'].isna()]
print("The boats which have not been reserved are:", [boat for boat in boats_not_reserved.index])

# Which client reserved the BLUE boat with the MOST reservations to date?

boats_clients.sort_values(by = 'n_reservations', ascending = False)
# Marie

The client who reserved 'Julia' is: Anna
The client who reserved 'Siren' is: Yann


The boats which have not been reserved are: ['Sea Sons', 'Cesar']


color  reservation_id  n_reservations  client_id client_name
boat_name                                                               
Hercules     blue               1              41       91.0       Marie
Julia        blue               2              34      154.0        Anna
Sea Sons      red               6              20        NaN         NaN
Minerva     green               5              16       22.0     Yassine
Cesar      yellow               4              12        NaN         NaN
Siren       green               3              10      124.0        Yann

## 4. Grouping the elements of a DataFrame: groupby, agg and crosstab methods.

    The groupby method allows you to group the rows of a DataFrame which share a common value on a given column.

    This method does not return a DataFrame. The object returned by the groupby method is an object of the DataFrameGroupBy class.

    This class is used to perform operations such as calculating statistics (sum, average, maximum, etc.) for each modality of the column on which the rows are grouped.

    The general structure of a groupby operation is as follows:

            Split the data.
            Apply a function.
            Combine the results.

    Example:

    It is assumed that the boats in the boats DataFrame are all identical and have the same age. We want to determine if the color of a boat has an influence on its number of reservations. For this, we will calculate for each color the average number of reservations per boat.

    It is therefore necessary to:

            Split the boats by color.
            Apply the mean function to compute the average number of reservations.
            Combine the results in a DataFrame to easily compare them.

    Therefore, we can use the groupby method followed by the mean method to get the result:

<img src="Photos\pd_groupby_en.png" width="900" height="500">


    All the usual statistical methods (count,mean, max, etc.) can be used as a suffix of the groupby method. These will only be applied on columns of compatible type.

    It is possible to specify for each column which function must be used in the Apply step of a groupby operation. For that, we use the agg method of the DataFrameGroupBy class by giving it a dictionary where each key is the name of a column and the value is the function to apply.

    Example:

    Let us go back to the transactions DataFrame:
    transaction_id 	cust_id 	tran_date 	prod_subcat_code 	prod_cat_code 	qty 	rate 	tax 	total_amt 	store_type
    80712190438 	270351 	28-02-2014 	1 	1 	-5 	-772 	405.3 	-4265.3 	e-Shop
    29258453508 	270384 	27-02-2014 	5 	3 	-5 	-1497 	785.925 	-8270.92 	e-Shop
    51750724947 	273420 	24-02-2014 	6 	5 	-2 	-791 	166.11 	-1748.11 	TeleShop
    93274880719 	271509 	24-02-2014 	11 	6 	-3 	-1363 	429.345 	-4518.35 	e-Shop
    51750724947 	273420 	23-02-2014 	6 	5 	-2 	-791 	166.11 	-1748.11 	TeleShop

    We want to determine, for each customer (cust_id), the minimum, maximum and the total amount spent from the total_amt column. We also want to know how many types of stores the customer has made a transaction in (store_type column).

    We can perform these calculations using a groupby operation:

            Split the transactions by the customer identifier.
            For the total_amt column, calculate the minimum (min), maximum (max) and the sum (sum). For the store_type column, count the number of unique modalities taken.
            Combine the results in a DataFrame.

    To find the number of unique modalities taken by the store_type column, we will use the following lambda function:

    import numpy as np

    n_modalities = lambda store_type: len(np.unique(store_type))

            The lambda function must take as argument a column and return a number.
            The np.unique function determines the unique modalities that appear in a sequence.
            The len function counts the number of elements in a sequence, i.e. its length.

    Thus, this function will allow us to determine the number of unique modalities for the store_type column.

    To apply these functions in the groupby operation, we'll use a dictionary whose keys are the columns to process and the values the functions to use.

    functions_to_apply = {
    # Classic statistical methods can be entered with
    # strings
    'total_amt': ['min', 'max', 'sum'],
    'store_type': n_modalities
    }

    This dictionary can now be fed into the agg method to perform the groupby operation:

    transactions.groupby('cust_id').agg(functions_to_apply)

    Which produces the following DataFrame:


                total_amt                               store_type

                 min               max          sum 	   <lambda>

    cust_id
    266783 	-5838.82 	5838.82 	3113.89 	2
    266784 	442 	4279.66 	5694.07 	3
    266785 	-6828.9 	6911.77 	21613.8 	3
    266788 	1312.74 	1927.12 	6092.97 	3
    266794 	-135,915 	4610.06 	27981.9 	4

    (a) Using a groupby operation, determine for each customer from the quantity of items purchased in a transaction (qty column):

            The maximum quantity.
            The minimum quantity
            The median quantity.

        You will have to filter the transactions whose quantity is negative. For this, you can use conditional indexing (qty[qty > 0]) over the column in a lambda function.



In [ ]:
# Insert your code here

func = {'qty' : ['max', 'min', 'median']}

transactions.loc[transactions.qty > 0].groupby('cust_id').agg(func).head()

# Original Cozum
# Maximal Quantity
max_qty = lambda qty: qty[qty > 0].max()

# Minimal Quantity
min_qty = lambda qty: qty[qty > 0].min()

# Median Quantity
median_qty = lambda qty : qty[qty > 0].median()

# Definition of the dictionnary of functions to apply
functions_to_apply = {
    'qty' : [max_qty, min_qty, median_qty]
}

# Groupby Operation
qty_groupby = transactions.groupby('cust_id').agg(functions_to_apply)

# For a better display, we can rename the columns produced by the groupby operation
qty_groupby.columns.set_levels(['max_qty', 'min_qty', 'median_qty'], level=1, inplace = True)

# Display of the first rows of the DataFrame produced by the groupby operation
qty_groupby.head()


In [ ]:


    Another way of grouping and summarizing data is to use the crosstab function of pandas which, as its name suggests, is used to crosstab the data in the columns of a DataFrame.

    A crosstab allows us to visualize the appearance frequency of pairs of modalities in a DataFrame.

    Example :

    In the transactions DataFrame, we want to know which are the most frequent category and subcategory pairs (prod_cat_code and prod_subcat_code columns)

    The crosstab function of pandas gives us this result:

    column1 = transactions['prod_cat_code']
    column2 = transactions['prod_subcat_code']
    pd.crosstab(column1, column2)

    This instruction produces the following DataFrame:
    prod_subcat_code

    prod_cat_code 	-1 	1 	2 	3 	4 	5 	6 	7 	8 	9 	10 	11 	12
    1 	4 	1001 	0 	981 	958 	0 	0 	0 	0 	0 	0 	0 	0
    2 	4 	934 	0 	1040 	1005 	0 	0 	0 	0 	0 	0 	0 	0
    3 	11 	0 	0 	0 	1020 	950 	0 	0 	966 	976 	945 	0 	0
    4 	5 	993 	0 	0 	988 	0 	0 	0 	0 	0 	0 	0 	0
    5 	3 	0 	0 	1023 	0 	0 	984 	1037 	0 	0 	998 	1029 	962
    6 	5 	0 	1002 	0 	0 	0 	0 	0 	0 	0 	1025 	1013 	1057

    The (i, j) cell of the resulting DataFrame contains the number of rows of the DataFrame having the modality i for column 1 and the modality j for column 2.

    Thus, it is easy to determine, for example, that the dominant subcategories of the category 4 are 1 and 4.

    The normalize argument of crosstab allows to display frequencies as a percentage.

    Thus, the argument normalize = 1 normalizes the table over the axis 1 of the crosstab, i.e. its columns:

    We recover the year of the transaction.
    column1 = transactions['tran_date'].apply(lambda x: x.split('-')[2]).astype(int)

    column2 = transactions[store_type]

    pd.crosstab(column1,
                column2,
                normalize = 1)

    This produces the following DataFrame:


    store_type

    tran_date 	Flagship store 	MBR 	TeleShop 	e-Shop
    2011 	0.291942 	0.323173 	0.283699 	0.306947
    2012 	0.331792 	0.322093 	0.336767 	0.322886
    2013 	0.335975 	0.3115 	0.332512 	0.320194
    2014 	0.0402906 	0.0432339 	0.0470219 	0.0499731

    This DataFrame allows us to say that 33.5975% of the transactions made in a 'Flagship store' took place in 2013.

    Conversely, by entering the argument normalize = 0, the crosstab is normalized over each row:


    store_type

    tran_date 	Flagship store 	MBR 	TeleShop 	e-Shop
    2011 	0.191121 	0.21548 	0.182617 	0.410781
    2012 	0.20096 	0.198693 	0.20056 	0.399787
    2013 	0.205522 	0.194074 	0.2 	0.400404
    2014 	0.173132 	0.189215 	0.198675 	0.438978

    Normalizing over the rows allows us to deduce that the transactions made in an 'e-Shop' account for 41.0781% of the transactions of the year 2011.

    In the covid_tests.csv file, we have a dataset of 200 COVID-19 tests. The columns of this dataset are as follows:

            'patient_id': ID of the patient tested.
            'test_result': Result of the test. Equals 1 if the patient is tested positive and 0 otherwise.
            'infected': Equals 1 if the patient was actually infected and 0 otherwise.

    (b) Load the dataset contained in the covid_tests.csv file. The values are separated by the character ';'.

    (c) Use the pd.crosstab function to determine the number of False Negatives produced by this test. (A false negative occurs when the test determines that the patient is not infected when they actually are.)

    (d) What is the false positive rate of the test? The false positive rate is the proportion of false positives in relation to the number of people that are not infected. (A false positive occurs when the test determines the patient is infected when they are not.)



In [ ]:
# Insert your code here
#b)
df = pd.read_csv('covid_tests.csv', sep=';')
df

#c)
col1 = df.test_result
col2 = df.infected

print(pd.crosstab(col1, col2))
print(30*'-', 'normalize = 1   3/3+71 ')
#d) normalize = 0

print(pd.crosstab(col1, col2, normalize=1))

# normalize = 1
print(30*'-', 'normalize = 0    3/3+119' )
print(pd.crosstab(col1, col2, normalize=0))


#Original Cozum
# Loading the dataset in 'covid_tests.csv'
covid_df = pd.read_csv("covid_tests.csv", sep = ';', index_col = 'patient_id')
covid_df.head()


# Crosstab of the test results with reality
pd.crosstab(covid_df['test_result'], 
            covid_df['infected'])

# There are 3 false negatives


pd.crosstab(covid_df['test_result'], 
            covid_df['infected'],
            normalize = 1)

# The false positive rate is about 5,6%
# 94,4% of healthy people are true negatives

In [ ]:
Conclusion and recap

    In this notebook you have learned to:

            Filter the rows of a DataFrame with multiple conditions using the binary operators &, | and -:

        # Year equal to 1979 and surface area greater than 60
        df[(df['year'] == 1979) & (df['surface'] > 60)]

        Year greater than 1900 or neighborhood equal to 'Père-Lachaise'.
        df[(df['year'] > 1900) | (df['neighborhood'] == 'Père-Lachaise')]

            Merge DataFrames using the concat function and the merge method.

        # Vertical concatenation
        pd.concat([df1, df2], axis = 0)

        # Horizontal concatenation
        pd.concat([df1, df2], axis = 1)

        # Different types of joins
        df1.merge(right = df2, on = 'column', how = 'inner')
        df1.merge(right = df2, on = 'column', how = 'outer')
        df1.merge(right = df2, on = 'column', how = 'left')
        df1.merge(right = df2, on = 'column', how = 'right')

            Sort and order the values of a DataFrame with the sort_values and sort_index methods.

        # Sorting a DataFrame by 'column' in ascending order
        df.sort_values(by = 'column', ascending = True)

            Perform a complex groupby operation using lambda functions and the groupby and agg methods.

        functions_to_apply = {
        'column1': ['min', 'max'],
        'column2' : [np.mean, np.std],
        'column3' : lambda x: x.max() - x.min()
        }

        df.groupby('column_to_group_by').agg(functions_to_apply)

    In this introductory module to Python for Data Science, you have learned how to create, clean and manipulate a dataset with Python using the numpy and pandas modules.

    You now have all the tools to approach more advanced Data Science notions such as Machine Learning or Data Visualization :)

